### ISSUE TITLE
Hotfix versioning does not conform to semantic versioning

### ISSUE THREAD

**Issue by buske** [MatchmakerExchange/mme-apis#112]
Currently, we state that versions take the form "vX.Y" in the header, but then we have hotfix versions of the form "v1.0a", which does not conform to [semantic versioning](http://semver.org/spec/v2.0.0.html). It is unclear if "v1.0" -> "v1.0" is a backward-compatible change or not, and currently it is not specified if the "a" goes in the header version specification or not.

**Comment by Relequestual** [comment]
I agree, we have left this a bit abigious. I was aware that the hotfix versioning I suggested didn't conform to semantic versioning methods, but considered it not that much of an issue, as we don't actually have any CODE in this repo which is imported via package managers to others codebases (yet).

Still, there's no reason NOT to do this properly.
I suggest for now that letters are stripped from version checks, and in future we use full semantic versioning as defined therein.

**Comment by Relequestual** [comment]
We could close the hotfix in a manual process rather than using automated gitflow. This would allow us to merge into master and develop, but then set the tag manually.

Also, we should remove the "v" from the versioning system. If we're going to do semantic versioning, may as well go all the way! This will require an update to our documentation, I think.

**Comment by fschiettecatte** [comment]
Just read the [semantic versioning] (http://semver.org/spec/v2.0.0.html) document, it makes perfect sense to adopt it all the way.

**Comment by Relequestual** [comment]
Agreed on technical call at time of writing, we will do this moving forward.

**Comment by Relequestual** [comment]
Based on the above, I think it's fair to consider hotfixes to be n.n.x where x is incremented, as opposed to x being alpha without a point break. +1's required for this, and I'll close it. Will be in effect from 1.1 onwards. 1.n.x where x is a non-breaking change as per Semantic Versioning.

**Comment by buske** [comment]
+1 We have to be okay with large version numbers if we fully commit to semver, and/or go back and revise some of the changes we've made since 1.0 to make them actually backwards-compatible.

**Comment by Relequestual** [comment]
I'm OK with large version numbers. Doing it properly!
Hum yeah that's a good point. I'll create a new issue for this and make it part of 1.1.

**Comment by Relequestual** [comment]
Summary:
We agree we should stick to Semantic Versioning.
We should remove the "v" from our version string

Is any system currently checking for version number and doing anything as a result of that check? DECIPHER is, but it wouldn't be difficult to account for v/no v. Thoughts?

**Comment by fschiettecatte** [comment]
GeneMatcher checks for versioning, currently it sends requests in the highest version it supports (but does not currently retrograde if the remote does not support it), and responds in the same version that was used in the request.

**Comment by harindra-a** [comment]
Hi,
As per spec, Broad checks (when live, currently internal dev) for
"application/vnd.ga4gh.matchmaker.v0.1+json" and returns a HTTP 406 if
there is no match,
Best,
Harindra

**Comment by Relequestual** [comment]
@harindra-a Great! We should probably all do the same based on the http status code spec.
We could (and should) discuss a means by which we can automatically downgrade the request to a supported version.

> The server SHOULD generate a payload containing a list of available representation characteristics and corresponding resource identifiers from which the user or user agent can choose the one most appropriate. A user agent MAY automatically select the most appropriate choice from that list. However, this specification does not define any standard for such automatic selection
https://httpstatuses.com/406

It would be great if we could build into the API, a means by which a system, as a 406 response, includes the supported versions.
I think this is linked to https://github.com/ga4gh/mme-apis/issues/116

**Comment by fschiettecatte** [comment]
GeneMatcher also generates a 406 with an error message which lists supported versions.

**Comment by buske** [comment]
I don't see any reason to bother removing the "v" in the Content-Type. The part after the "v" should just be the semver, and probably just the first two digits (e.g., https://github.com/ga4gh/server/issues/526). The API itself can be tagged/released with three-digit precision if we'd like.

I want to clarify our approach to version negotiation because our API isn't quite like normal content negotiation and our different groups haven't been uniform in the way we handle the `Accept` and `Content-Type` headers.

API versioning was designed to be handled like any other content negotiation: The `Accept` header should include the content types of one or more API versions that the client supports, and the server should return with a supported `Content-Type` or return an error with a list of supported versions. However, it has become clear that our API doesn't work quite like this: unlike normal content negotiation, the query is also in a `Content Type` that might or might not be understood by the server. In other words, this is a dual-negotiation process, and therefore really both the client and server need to be speaking a compatible version from the start.

As an example for why normal content negotiation doesn't make sense: If the client supports v1 and v2, it could send a request on the v2 API with an `Accept: v2; v1` and a `Content-Type: v2`. If the server only accepts v1, it's not possible for the server to understand the request and return results using v1 immediately, which is what content negotiation is designed to do (in my understanding). Instead, the server needs to reply with "406: Sorry, I don't speak v2, but I do speak: v1". The client then has to re-issue the query on v1. The `Accept` header was irrelevant for this negotiation.

As such, I would propose:
- we switch to only using the `Content-Type` header
- we tag github releases with 3-digit precision semver, where:
  - the last digit (patches) are used for documentation, test data, infrastructure changes, but don't affect the API proper
  - the middle digit is for backwards-compatible API changes, which basically only corresponds to the addition of new optional fields or optional endpoints. Even changing required fields to be optional or vise versa breaks backwards compatibility
  - the first digit is for non-backwards-compatible changes, which include the changing whether fields are optional/required, the addition of required fields, the structure of fields or values, etc.

Thoughts?

**Comment by Relequestual** [comment]
Agreed on all of the above. I think when we work this change into our documentation, we should include a diagram along with the text example. I had to read it a few times, but that might just be lack of coffee today.

semver all the way!

**Comment by Relequestual** [comment]
This was briefly discussed on one of our MME technical calls in the last month or so. I mentioned that although contenet negotiation as a principal is nothing new, we may want to suggest ot give examples as to how to actuall DO it as part of the docs.
